src/topology/perfect.lean

```diff
@@ -0,0 +1,231 @@
+  preperfect (U ∩ C) :=
+begin
+  rintros x ⟨xU, xC⟩,
+  apply (hC _ xC).nhd_inter,
+  exact hU.mem_nhds xU,
+end
+
+/-- The closure of a preperfect set is perfect.
+For a converse, see `preperfect_iff_perfect_closure`-/
+theorem preperfect.perfect_closure (hC : preperfect C) : perfect (closure C) :=
+begin
+  split, { exact is_closed_closure },
+  intros x hx,
+  by_cases h : x ∈ C; apply acc_pt.mono _ (principal_mono.mpr subset_closure),
+  { exact hC _ h },
+  have : {x}ᶜ ∩ C = C := by simp[h],
+  rw [acc_pt, nhds_within, inf_assoc, inf_principal, this],
+  rw [closure_eq_cluster_pts] at hx,
+  exact hx,
+end
+
+/-- In a T1 space, being preperfect is equivalent to having perfect closure.-/
+theorem preperfect_iff_perfect_closure [t1_space α] :
+  preperfect C ↔ perfect (closure C) :=
+begin
+  split; intro h, { exact h.perfect_closure },
+  intros x xC,
+  have H := h.acc _ (subset_closure xC),
+  rw acc_pt_iff_frequently at *,
+  have : ∀ y , y ≠ x ∧ y ∈ closure C → ∃ᶠ z in 𝓝 y, z ≠ x ∧ z ∈ C,
+  { rintros y ⟨hyx, yC⟩,
+    simp only [← mem_compl_singleton_iff, @and_comm _ (_ ∈ C) , ← frequently_nhds_within_iff,
+      hyx.nhds_within_compl_singleton, ← mem_closure_iff_frequently],
+    exact yC, },
+  rw ← frequently_frequently_nhds,
+  exact H.mono this,
+end
+
+theorem perfect.closure_nhd_inter {U : set α} (hC : perfect C) (x : α) (xC : x ∈ C) (xU : x ∈ U)
+  (Uop : is_open U) : perfect (closure (U ∩ C)) ∧ (closure (U ∩ C)).nonempty :=
+begin
+  split,
+  { apply preperfect.perfect_closure,
+    exact (hC.acc).open_inter Uop, },
+  apply nonempty.mono subset_closure,
+  exact ⟨x, ⟨xU, xC⟩⟩,
+end
+
+/-- Given a perfect nonempty set in a T2.5 space, we can find two disjoint perfect subsets
+This is the main inductive step in the proof of the Cantor-Bendixson Theorem-/
+lemma perf_nonempty.splitting [t2_5_space α] (hC : perfect C) (hnonempty : C.nonempty) :
+  ∃ C₀ C₁ : set α, (perfect C₀ ∧ C₀.nonempty ∧ C₀ ⊆ C) ∧
+  (perfect C₁ ∧ C₁.nonempty ∧ C₁ ⊆ C) ∧ disjoint C₀ C₁ :=
+begin
+  cases hnonempty with y yC,
+  obtain ⟨x, xC, hxy⟩ : ∃ x ∈ C, x ≠ y,
+  { have := hC.acc _ yC,
+    rw acc_pt_iff_nhds at this,
+    rcases this univ (univ_mem) with ⟨x,xC,hxy⟩,
+    exact ⟨x,xC.2,hxy⟩, },
+  obtain ⟨U, xU, Uop, V, yV, Vop, hUV⟩ := exists_open_nhds_disjoint_closure hxy,
+  use [closure (U ∩ C), closure (V ∩ C)],
+  split; rw ← and_assoc,
+  { refine ⟨hC.closure_nhd_inter x xC xU Uop, _⟩,
+    rw hC.closed.closure_subset_iff,
+    apply inter_subset_right, },
+  split,
+  { refine ⟨hC.closure_nhd_inter y yC yV Vop, _⟩,
+    rw hC.closed.closure_subset_iff,
+    apply inter_subset_right, },
+  apply disjoint.mono _ _ hUV; apply closure_mono; apply inter_subset_left,
+end
+
+section kernel
+
+/-- The Cantor-Bendixson Theorem: Any closed subset of a second countable space
+can be written as the union of a countable set and a perfect set.-/
+theorem countble_union_perfect_of_closed [second_countable_topology α]
```

Comment:
Name suggestion:
```suggestion
theorem exists_countable_union_perfect_of_is_closed [second_countable_topology α]
```
Design suggestion: if this is an important result as the docstring indicates, and you intend to take countable and perfect parts of a set often, you should add two definitions, say `is_closed.countable_part` and `is_closed.perfect_part`, that give you two such sets chosen arbitrarily (with `Exists.some`/`classical.some`), and add a lemma stating that their union is `C`. Doing this will avoid needing to unpack the exists every time you want to use that lemma.

The next lemma (currently named `perf_nonempty_of_closed_unctble`) can also be rephrased as a lemma about `is_closed.perfect_part`, which proves that it is nonempty when the set `C` is not countable.